crates/bevy_ecs/src/query/iter.rs

```diff
@@ -50,6 +50,37 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIter<'w, 's, D, F> {
         }
     }
 
+    /// Reborrows this iterator, creating a new iterator that will yield the same items but
+    /// won't advance the original iterator, thus later allowing to restart the iteration from
+    /// the current point. The original iterator will be unusable while the new one exists.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_ecs::prelude::*;
+    /// #
+    /// # #[derive(Component)]
+    /// # struct ComponentA;
+    ///
+    /// fn combinations(mut query: Query<&mut ComponentA>) {
+    ///     let mut iter = query.iter_mut();
+    ///     while let Some(a) = iter.next() {
+    ///         for b in iter.reborrow() {
+    ///             // Check every combination (a, b)
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    pub fn reborrow(&mut self) -> QueryIter<'_, 's, D, F> {
```

Comment:
I decided to add both a `remainder` and a `remainder_mut` methods since I've noticed this feature was not available for read-only iterators, and mutably borrowing the iterator might be limiting for those cases.